src/statsd-parsers/basic/basic.c

```diff
@@ -3,130 +3,179 @@
 #include <stdint.h>
 #include <string.h>
 #include <sys/types.h>
-#include "../statsd-parsers.h"
 #include "../../utils/utils.h"
+#include "../statsd-parsers.h"
+#include "basic.h"
 
 #define JSON_BUFFER_SIZE 4096
 
-statsd_datagram* basic_parser_parse(char* buffer) {
-    struct statsd_datagram* datagram = (struct statsd_datagram*) malloc(sizeof(struct statsd_datagram));
-    *datagram = (struct statsd_datagram) {0};
+int basic_parser_parse(char* buffer, statsd_datagram** datagram) {
+    *datagram = (struct statsd_datagram*) malloc(sizeof(struct statsd_datagram));
+    *(*datagram) = (struct statsd_datagram) {0};
+    if (parse(buffer, datagram)) {
+        buffer[strcspn(buffer, "\n")] = 0;
+        verbose_log("Parsed: %s", buffer);
+        return 1;
+    }
+    free(*datagram);
+    buffer[strcspn(buffer, "\n")] = 0;
+    verbose_log("Thrown away. REASON: unable to parse: %s", buffer);
+    return 0;
+};
+
+int parse(char* buffer, statsd_datagram** datagram) {
     int current_segment_length = 0;
     int i = 0;
     char previous_delimiter = ' ';
     int count = strlen(buffer);
-    char *segment = (char *) malloc(count); // cannot overflow since whole segment is count anyway
+    char* segment = (char *) malloc(count); // cannot overflow since whole segment is count anyway
     ALLOC_CHECK("Unable to assign memory for StatsD datagram message parsing.");
     const char INSTANCE_TAG_IDENTIFIER[] = "instance";
-    char *json_key = NULL;
-    char *json_value = NULL;
+    char* tag_key = NULL;
+    char* tag_value = NULL;
+    char* attr;
     char tags_json_buffer[JSON_BUFFER_SIZE];
     memset(tags_json_buffer, '\0', JSON_BUFFER_SIZE);
```

Comment:
Looks like `tags_json_buffer` is unused now. But as I wrote, I actually like the idea of storing all tags in a JSON string, we don't need them separately there will be no processing at all - if they are in some expected order (e.g. alphabetical) then they can be used for hashing input as a string and then handed over into PCP easily as this is the expected format.

Reply:
Waiting for your feedback to https://github.com/Erbenos/pcp-statsd-c/pull/5#discussion_r290398541

Reply:
It's up to you how you want to store it, however if the agent will not provide ability to list metrics by labels, I think it's easier just to store it in the final form as soon as possible.